huaweicloud/services/cbr/resource_huaweicloud_cbr_vault.go

```diff
@@ -0,0 +1,497 @@
+	if len(resources) > 1 {
+		return []vaults.ResourceCreate{}, fmtp.Errorf("[ERROR] The size of resources cannot grant than 1 for disk and turbo vault")
+	} else if len(resources) == 0 {
+		return []vaults.ResourceCreate{}, nil
+	}
+	res := resources[0].(map[string]interface{})
+	if res["includes"].(*schema.Set).Len() > 0 {
+		result := make([]vaults.ResourceCreate, res["includes"].(*schema.Set).Len())
+		for i, v := range res["includes"].(*schema.Set).List() {
+			result[i] = vaults.ResourceCreate{
+				ID:   v.(string),
+				Type: rType,
+			}
+		}
+		return result, nil
+	}
+	return []vaults.ResourceCreate{}, fmtp.Errorf("[ERROR] Only includes can be set for disk and turbo vault")
+}
+
+func buildCBRVaultResources(vType string, resources *schema.Set) ([]vaults.ResourceCreate, error) {
+	var result []vaults.ResourceCreate
+	var err error
+	rType, ok := ResourceType[vType]
+	if !ok {
+		fmtp.Errorf("[ERROR] Only includes can be set for disk and turbo vault")
+	}
+	logp.Printf("[DEBUG] The resource type is %s", rType)
+	switch rType {
+	case ResourceTypeServer:
+		result, err = buildCbrVaultResourcesForServer(rType, resources.List())
+	case ResourceTypeDisk, ResourceTypeTurbo:
+		result, err = buildCbrVaultResourcesForDisk(rType, resources.List())
+	default:
+		err = fmtp.Errorf("[ERROR] Resource type does not support: %s", rType)
+	}
+	return result, err
+}
+
+func buildCBRVaultBilling(d *schema.ResourceData) *vaults.BillingCreate {
+	billing := &vaults.BillingCreate{
+		ObjectType:      d.Get("type").(string),
+		ConsistentLevel: d.Get("consistent_level").(string),
+		ProtectType:     d.Get("protection_type").(string),
+		Size:            d.Get("size").(int),
+	}
+
+	return billing
+}
+
+func resourceCBRVaultV3Create(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	client, err := config.CbrV3Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.Errorf("Error creating Huaweicloud CBR v3 client: %s", err)
+	}
+
+	resources, err := buildCBRVaultResources(d.Get("type").(string), d.Get("resources").(*schema.Set))
+	if err != nil {
+		return fmtp.Errorf("Error building vault resources: %s", err)
+	}
+	opts := vaults.CreateOpts{
+		Name:                d.Get("name").(string),
+		AutoExpand:          d.Get("auto_expand").(bool),
+		BackupPolicyID:      d.Get("policy_id").(string),
+		EnterpriseProjectID: config.GetEnterpriseProjectID(d),
+		Resources:           resources,
+		Billing:             buildCBRVaultBilling(d),
+	}
+
+	logp.Printf("[DEBUG] The createOpts is: %+v", opts)
+	vault, err := vaults.Create(client, opts).Extract()
+	if err != nil {
+		return fmtp.Errorf("Error creating vaults: %s", err)
+	}
+	d.SetId(vault.ID)
+
+	if policy, ok := d.GetOk("policy_id"); ok {
+		_, err := vaults.BindPolicy(client, d.Id(), vaults.BindPolicyOpts{PolicyID: policy.(string)}).Extract()
+		if err != nil {
+			return fmtp.Errorf("Error binding policy to vault: %s", err)
+		}
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
```

Comment:
The tags package already provides a common method to create tags, which can be used later.

Reply:
done